gallery/gallery.py

```diff
@@ -0,0 +1,205 @@
+        return f"https://github.com/python/cpython/archive/{version}.zip"
+    elif package == "pypy":
+        if version is None:
+            version = "branch/default"
+        return (
+            f"https://foss.heptapod.net/pypy/pypy/repository/{version}/archive.tar.bz2"
+        )
+    else:
+        return get_pypi_download_url(package, version)
+
+
+def get_archive_manager(local_file: str) -> ArchiveKind:
+    if tarfile.is_tarfile(local_file):
+        return tarfile.open(local_file)
+    elif zipfile.is_zipfile(local_file):
+        return zipfile.ZipFile(local_file)
+    else:
+        raise ValueError("Unknown archive kind.")
+
+
+def get_first_archive_member(archive: ArchiveKind) -> str:
+    if isinstance(archive, tarfile.TarFile):
+        return archive.getnames()[0]
+    elif isinstance(archive, zipfile.ZipFile):
+        return archive.namelist()[0]
+
+
+def download_and_extract(package: str, version: Optional[str], directory: Path) -> Path:
+    source = get_package_source(package, version)
+
+    local_file, _ = urlretrieve(source, directory / f"{package}-src")
+    with get_archive_manager(local_file) as archive:
+        archive.extractall(path=directory)
+        result_dir = get_first_archive_member(archive)
+    return directory / result_dir
+
+
+def git_create_repository(repo: Path) -> None:
+    subprocess.run(["git", "init"], cwd=repo)
+    git_add_and_commit(msg="Inital commit", repo=repo)
+
+
+def git_add_and_commit(msg: str, repo: Path) -> None:
+    subprocess.run(["git", "add", "."], cwd=repo)
+    subprocess.run(["git", "commit", "-m", msg, "--allow-empty"], cwd=repo)
+
+
+def git_switch_branch(
+    branch: str, repo: Path, new: bool = False, from_branch: Optional[str] = None
+) -> None:
+    args = ["git", "checkout"]
+    if new:
+        args.append("-b")
+    args.append(branch)
+    if from_branch:
+        args.append(from_branch)
+    subprocess.run(args, cwd=repo)
+
+
+def init_repo(options: Namespace) -> Path:
+    options.output.mkdir(exist_ok=True)
+
+    source_directory = download_and_extract(
+        package=options.pypi_package, version=options.version, directory=options.output
+    )
+    git_create_repository(source_directory)
+    if options.black_repo is None:
+        subprocess.run(
+            ["git", "clone", "https://github.com/psf/black.git", INTERNAL_BLACK_REPO],
+            cwd=options.output,
+        )
+        subprocess.run(
+            ["pip", "install", "-e", INTERNAL_BLACK_REPO], cwd=options.output
+        )
+        options.black_repo = options.output / INTERNAL_BLACK_REPO
+    else:
+        subprocess.run(["pip", "install", "black"])
+
+    return source_directory
+
+
+def format_repo_with_version(
+    repo: Path,
+    from_branch: Optional[str],
+    black_repo: Path,
+    black_version: BlackVersion,
+    input_directory: Path,
+) -> str:
+    current_branch = f"black-{black_version.version}"
+    git_switch_branch(black_version.version, repo=black_repo)
```

Comment:
It would be better to have `pip install -e .` after switching branches. Different branches are bound to have different dependencies. So, ideally, for reproducibility, we'd use different virtual environments for each install, too.